packages/core/src/new-api/internal/execution/transactions/transaction-service.ts

```diff
@@ -0,0 +1,69 @@
+import { IgnitionError } from "../../../../errors";
+import { ArtifactResolver } from "../../../types/artifact";
+import {
+  DeployContractInteraction,
+  DeployContractResult,
+  OnchainInteraction,
+  OnchainResult,
+} from "../../../types/journal";
+import { TransactionService } from "../../../types/transaction-service";
+import { isDeployContractInteraction } from "../guards";
+
+import { ChainDispatcher } from "./chain-dispatcher";
+
+/**
+ * A service for managing on-chain interactions during the deployment.
+ *
+ * @beta
+ */
+export class TransactionServiceImplementation implements TransactionService {
+  constructor(
+    private _artifactLoader: ArtifactResolver,
+    private _chainDispatcher: ChainDispatcher
+  ) {}
+
+  public async onchain(
```

Comment:
I think the missing part I don't fully grasp is how this would work with things like re-sending a transaction because it was dropped, bumping gas, etc.